server/src/main/java/org/opensearch/index/store/Store.java

```diff
@@ -700,6 +700,49 @@ public void cleanupAndVerify(String reason, MetadataSnapshot sourceMetadata) thr
         }
     }
 
+    /**
+     * This method deletes every file in this store that is not contained in either the remote or local metadata snapshots.
+     * This method is used for segment replication when the in memory SegmentInfos can be ahead of the on disk segment file.
+     * In this case files from both snapshots must be preserved.
+     * @param reason         the reason for this cleanup operation logged for each deleted file
+     * @param remoteSnapshot The remote snapshot sent from primary shards.
+     * @param localSnapshot  The local snapshot from in memory SegmentInfos.
+     * @throws IllegalStateException if the latest snapshot in this store differs from the given one after the cleanup.
+     */
+    public void cleanupAndVerify(String reason, MetadataSnapshot remoteSnapshot, MetadataSnapshot localSnapshot) throws IOException {
+        // fetch a snapshot from the latest on disk Segments_N file. This can be behind
+        // the passed in local in memory snapshot, so we want to ensure files it references are not removed.
+        final Store.MetadataSnapshot latestCommitPointMetadata = getMetadata((IndexCommit) null);
+        metadataLock.writeLock().lock();
+        try (Lock writeLock = directory.obtainLock(IndexWriter.WRITE_LOCK_NAME)) {
+            for (String existingFile : directory.listAll()) {
+                if (Store.isAutogenerated(existingFile)
+                    || remoteSnapshot.contains(existingFile)
+                    || latestCommitPointMetadata.contains(existingFile)) {
```

Comment:
Most of the code in this method is identical to the existing `cleanupAndVerify` function. I would suggest reworking this to either:

1. A method that has hooks for the if-condition check and the cleanup steps, and passing in the relevant logic for segrep vs docrep

OR

2. Extracting the cleanup logic to its own class, with hooks for the if-condition and cleanup steps, and having the segrep implementation be a subclass